### ISSUE TITLE
Nothings happens when setting 'withInternet: true'

### ISSUE THREAD

**Issue by rdmo94** [flutternetwork/WiFiFlutter#235]
I'm using this library to connect to a Raspberry Pi connected to a wifi-network. I recently upgraded the device i'm testing on, and I am currently running android 11.

The call to `connect()` worked fine until I upgraded to a newer device and android version. Whenever i set the `withInternet` flag to false, it connects as intended, but as a temporary connection without internet access.

If i set `withInternet: true` nothing happens, but the method `connect` returns true, as if I'm connected successfully. 
I'm using WPA for security.

I should have the required permission set in my AndroidManifest, but i might be missing something.  Can anyone help?

**Comment by daadu** [comment]
There are two different API for newer version of Android. For connection with internet you can look for "Wifi Suggestion API". You can check it's docs.

Also I've tested this API for sometime, might review it in some time. Until, then encourage you to check the docs and code.

**Comment by rdmo94** [comment]
Only when withInternet is set to false. Otherwise I see no dialogue or anything.

**Comment by daadu** [comment]
@rdmo94 Sounds like an issue, will look into it this weekend (no promise).

**Comment by rdmo94** [comment]
Let me know if you need any further details!

**Comment by manu-hr** [comment]
Hi, even I’m having the same issue, if I use connect method  with  withInternet: true I’m getting a response
as " E/WifiIotPlugin( 3997): status: 0 " and with the method findAndConnect getting "PlatformException(Error, Invalid BSSID representation, , null)". 
This is with android 11 but android 8 and 9 are working fine.

**Comment by rdmo94** [comment]
Additionally I also experienced this error quite some times: 
java.lang.IllegalArgumentException: NetworkCallback was not registered, with a stack trace related to WifiIotPlugin. I'm not sure why and when it occurs, but it could be due to wifi being turned off.

**Comment by rdmo94** [comment]
![Screenshot 2022-01-25 at 12 13 55](https://user-images.githubusercontent.com/47812442/150967014-be8a8aac-5c23-48cb-8040-28f19a03d95d.png)

For reference

**Comment by manu-hr** [comment]
> Hi, even I’m having the same issue, if I use connect method with withInternet: true I’m getting a response as " E/WifiIotPlugin( 3997): status: 0 " and with the method findAndConnect getting "PlatformException(Error, Invalid BSSID representation, , null)". This is with android 11 but with android 8 and 9 working fine.

@daadu As for findAndConnect method the ssid and bssid parameters are interchanged, please look into it.

Even after that with WifiNetworkSuggestion API, I'm getting the connection status "STATUS_NETWORK_SUGGESTIONS_SUCCESS" but Nothing changing in the wifi connection of the device.
I'm getting the response also as true without waiting any time and no dialogues are popping up. If you have any suggestions please let me know.

**Comment by daadu** [comment]
I did some debugging, looks like the network added via the `NetworkSuggestion API` [which is the only way to get "internet enabled" wifi connection ] is just a "suggestion" to android - I found out that the android connects to the wifi - if you open the Wifi pick list from notification drawer or settings - looks like the android needs to be "triggered" to check the nely added suggestion (looks like a bug in Android side) - one workaround I found is to call `await WiFiForIoTPlugin.loadWifiList()` after `connect(..., withInternet: true)` - this will trigger android to look for app suggested network - and the connection is established.

@manu-hr Thanks for reporting the mac-adress issue, it is fixed with #245.

**Comment by daadu** [comment]
@rdmo94 @manu-hr Try the workaround I suggested and let me know if it works - will add this caveat in "README".

**Comment by manu-hr** [comment]
> @rdmo94 @manu-hr Try the workaround I suggested and let me know if it works - will add this caveat in "README".

No, It is not working for me. But as you mentioned if I open the wifi picklist it is connecting only if there is no other available wifi -connections.

**Comment by daadu** [comment]
Seems, like things are not in our hands.

Did you try `registerWifiNetwork` - see if that fits your use-case. The API is like user adding this network to the wifi list himself.

**Comment by daadu** [comment]
Moreover, I have prototyped removing "Network Suggestion API" at all - some SO answeres suggested using the "Network Specifier API" for internet connection as well.

You can test this against - [ daadu:wifi_iot/poc/remove-wifi-suggestion-api](https://github.com/daadu/WiFiFlutter/tree/wifi_iot/poc/remove-wifi-suggestion-api) + `forceWifiUsage` - see it that works.

**Comment by rdmo94** [comment]
@daadu Thanks a lot for your effort. I tried your suggested workaround with `await WiFiForIoTPlugin.loadWifiList()` without any change. Registering the network before connecting has no effect either.  Still nothing happens when trying to connect. As this is to be used in a business context, a colleague of mine will put some effort into fixing this. I'll let you know, if we come up with a permanent solution.

**Comment by Nailer** [comment]
Have the same issue here. 

When withInternet set to false I get the special native connection dialog that allows me to connect to my raspberry pi access point. 

However, when connected without internet, the Web-RTC negotiation via websocket is blocked. Somehow Android thinks we "need internet" to access local addresses like 192.168.4.1 etc.

When I set withInternet to true, the connect call returns success, but isn't actually connected.

**Comment by daadu** [comment]
@Nailer did you use `forceWifiUsage(true)` after connection? That might be the issue with `withInternet: false`.

With `withInternet: true` there seems to be an issue, which is under investigation.

**Comment by Nailer** [comment]
I did not try that. I will try it and see if it helps.

**Comment by quoc-huynh-cosee** [comment]
Any updates on this?

**Comment by rdmo94** [comment]
The issue still persists. In my project we've had to built our own native tool for newer Android versions using the `WifiNetworkSuggestion.Builder`. Using this approach Android will present a list of wifi suggestions, as the user has to manually select the wifi to connect to. Not sure why they changed this, but it appears there is no other way if you want internet connection.

**Comment by quoc-huynh-cosee** [comment]
> The issue still persists. In my project we've had to built our own native tool for newer Android versions using the `WifiNetworkSuggestion.Builder`. Using this approach Android will present a list of wifi suggestions, as the user has to manually select the wifi to connect to. Not sure why they changed this, but it appears there is no other way if you want internet connection.

Can you give me some references?

**Comment by rdmo94** [comment]
https://developer.android.com/reference/android/net/wifi/WifiNetworkSuggestion.Builder here is a reference to the Android API used. 

`private fun connectToWifi(ssid: String?, password: String?): Boolean {
        if(ssid == null || password == null) return false;

        val suggestion = WifiNetworkSuggestion.Builder()
            .setSsid(ssid)
            .setWpa2Passphrase(password)
            .build();

        val suggestions = listOf(suggestion);

        val wifiManager = context.getSystemService(Context.WIFI_SERVICE) as WifiManager;
        val status = wifiManager.addNetworkSuggestions(suggestions);

        val panelIntent = Intent(Panel.ACTION_WIFI)
        startActivityForResult(panelIntent, 545)

        return status == WifiManager.STATUS_NETWORK_SUGGESTIONS_SUCCESS
    }
}` 

Here is the code for the function we call from flutter. It's written in kotlin. 
You would have to create a channel for calling the function from flutter. Here is some flutter documentation on that:  https://docs.flutter.dev/development/platform-integration/platform-channels

Basically I reference this function through a named channel, which is pretty straight forward. The reference from flutter looks like this: 
`await _channel.invokeMethod(
          "connect", {"ssid": ssid, "password": password});`

**Comment by jassim18** [comment]
> https://developer.android.com/reference/android/net/wifi/WifiNetworkSuggestion.Builder here is a reference to the Android API used.
> 
> ```
> private fun connectToWifi(ssid: String?, password: String?): Boolean {
>         if(ssid == null || password == null) return false;
>         val suggestion = WifiNetworkSuggestion.Builder()
>             .setSsid(ssid)
>             .setWpa2Passphrase(password)
>             .build();
> 
>         val suggestions = listOf(suggestion);
>         val wifiManager = context.getSystemService(Context.WIFI_SERVICE) as WifiManager;
>         val status = wifiManager.addNetworkSuggestions(suggestions);
>         val panelIntent = Intent(Panel.ACTION_WIFI)
>         startActivityForResult(panelIntent, 545)
>         return status == WifiManager.STATUS_NETWORK_SUGGESTIONS_SUCCESS
> }
> ```
> 
> Here is the code for the function we call from flutter. It's written in kotlin. You would have to create a channel for calling the function from flutter. Here is some flutter documentation on that: https://docs.flutter.dev/development/platform-integration/platform-channels
> 
> Basically I reference this function through a named channel, which is pretty straight forward. The reference from flutter looks like this: `await _channel.invokeMethod( "connect", {"ssid": ssid, "password": password});`

Is there any way to disconnect? @rdmo94

**Comment by rdmo94** [comment]
> > https://developer.android.com/reference/android/net/wifi/WifiNetworkSuggestion.Builder here is a reference to the Android API used.
> > ```
> > private fun connectToWifi(ssid: String?, password: String?): Boolean {
> >         if(ssid == null || password == null) return false;
> >         val suggestion = WifiNetworkSuggestion.Builder()
> >             .setSsid(ssid)
> >             .setWpa2Passphrase(password)
> >             .build();
> > 
> >         val suggestions = listOf(suggestion);
> >         val wifiManager = context.getSystemService(Context.WIFI_SERVICE) as WifiManager;
> >         val status = wifiManager.addNetworkSuggestions(suggestions);
> >         val panelIntent = Intent(Panel.ACTION_WIFI)
> >         startActivityForResult(panelIntent, 545)
> >         return status == WifiManager.STATUS_NETWORK_SUGGESTIONS_SUCCESS
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Here is the code for the function we call from flutter. It's written in kotlin. You would have to create a channel for calling the function from flutter. Here is some flutter documentation on that: https://docs.flutter.dev/development/platform-integration/platform-channels
> > Basically I reference this function through a named channel, which is pretty straight forward. The reference from flutter looks like this: `await _channel.invokeMethod( "connect", {"ssid": ssid, "password": password});`
> 
> Is there any way to disconnect? @rdmo94

`private fun disconnectFromWifi(): Boolean {
        var wifiManager: WifiManager = context.getSystemService(Context.WIFI_SERVICE) as WifiManager;
        var status = wifiManager.removeNetworkSuggestions(mutableListOf<WifiNetworkSuggestion>());
        if(status == WifiManager.STATUS_NETWORK_SUGGESTIONS_SUCCESS) {
            return true;
        } else {
            return false;
        }

    }
`
This is how i do it. And i have a channel for calling the function.

**Comment by jassim18** [comment]
> > > https://developer.android.com/reference/android/net/wifi/WifiNetworkSuggestion.Builder here is a reference to the Android API used.
> > > ```
> > > private fun connectToWifi(ssid: String?, password: String?): Boolean {
> > >         if(ssid == null || password == null) return false;
> > >         val suggestion = WifiNetworkSuggestion.Builder()
> > >             .setSsid(ssid)
> > >             .setWpa2Passphrase(password)
> > >             .build();
> > > 
> > >         val suggestions = listOf(suggestion);
> > >         val wifiManager = context.getSystemService(Context.WIFI_SERVICE) as WifiManager;
> > >         val status = wifiManager.addNetworkSuggestions(suggestions);
> > >         val panelIntent = Intent(Panel.ACTION_WIFI)
> > >         startActivityForResult(panelIntent, 545)
> > >         return status == WifiManager.STATUS_NETWORK_SUGGESTIONS_SUCCESS
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Here is the code for the function we call from flutter. It's written in kotlin. You would have to create a channel for calling the function from flutter. Here is some flutter documentation on that: https://docs.flutter.dev/development/platform-integration/platform-channels
> > > Basically I reference this function through a named channel, which is pretty straight forward. The reference from flutter looks like this: `await _channel.invokeMethod( "connect", {"ssid": ssid, "password": password});`
> > 
> > Is there any way to disconnect? @rdmo94
> 
> ```
> private fun disconnectFromWifi(): Boolean {
>         var wifiManager: WifiManager = context.getSystemService(Context.WIFI_SERVICE) as WifiManager;
>         var status = wifiManager.removeNetworkSuggestions(mutableListOf<WifiNetworkSuggestion>());
>         if(status == WifiManager.STATUS_NETWORK_SUGGESTIONS_SUCCESS) {
>             return true;
>         } else {
>             return false;
>         }
>     }
> ```
> This is how i do it. And i have a channel for calling the function.

This doesn't disconnect the network. Unfortunately, we have to manually switch on and off the wifi to get disconnected from the network. Is there any way to override this manual process using program?

**Comment by jassim18** [comment]
Can anyone help me using example to remove network suggestions along with disconnect ? In documentation, it's mentioned that we should use constants. But it seems unfamiliar. It would be helpful if any example code is posted .

**Comment by daadu** [comment]
@jassim18 with the plugin - if the network was connected (network suggestion added) with plugin then you can remove them with `disconnect`

**Comment by jassim18** [comment]
> @jassim18 with the plugin - if the network was connected (network suggestion added) with plugin then you can remove them with `disconnect`

Hey @daadu . Thanks for checking out. But I need a working code example. I couldn't establish disconnect in my project.

**Comment by tmtong** [comment]
Any update on the connect issue?
I have the exact same problem. If useinternet is false, it have a dialog to connect. But use internet is true, it return true, and nothing happen, no dialog, no change in connection

**Comment by daadu** [comment]
@tmtong @jassim18 I am afraid I wont be able to give time on this issue as of now, I am a bit busy with other work. Feel free to debug it on your own and write a PR with the fix.

**Comment by tmtong** [comment]
I ended up having my own kotlin file. 

```
package com.example.ppap_console

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.net.wifi.WifiNetworkSpecifier
import android.net.wifi.WifiNetworkSuggestion
import android.net.wifi.WifiManager
import android.provider.Settings.Panel
import androidx.annotation.NonNull
import io.flutter.embedding.android.FlutterActivity
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel
import 	android.content.Intent
import android.content.IntentFilter
import android.content.BroadcastReceiver

class MainActivity : FlutterActivity() {
  private val CHANNEL = "YOURCHANNELNAME"

  override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
    super.configureFlutterEngine(flutterEngine)
    MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL).setMethodCallHandler {
        call,
        result ->
      if (call.method == "connect") {
        val temp = connectToWifiB(call.argument<String>("ssid").toString(), call.argument<String>("password").toString())
        result.success(temp)
        
      }
    }
  }
  // Wi-Fi Network Request API for peer-to-peer connectivity
  // The resulting network is not intended to provide internet access, 
  // can't be used by the system, and 
  // can't be used by any app except the configuring app.
  
  private fun connectToWifiA(ssid: String, password: String): Boolean {
    
      val wifiNetworkSpecifier: WifiNetworkSpecifier?

      wifiNetworkSpecifier = WifiNetworkSpecifier.Builder().setSsid(ssid)
              .setWpa2Passphrase(password).build()
      

      val networkRequest = NetworkRequest.Builder()
          .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
          .setNetworkSpecifier(wifiNetworkSpecifier)
          .build()

      val connectivityManager =
          context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager?

      val networkCallback = object : ConnectivityManager.NetworkCallback() {
          override fun onUnavailable() {
              super.onUnavailable()
          }

          override fun onLosing(network: Network, maxMsToLive: Int) {
              super.onLosing(network, maxMsToLive)

          }

          override fun onAvailable(network: Network) {
              super.onAvailable(network)
              connectivityManager?.bindProcessToNetwork(network)
          }

          override fun onLost(network: Network) {
              super.onLost(network)

          }
      }
      connectivityManager?.requestNetwork(networkRequest, networkCallback)
      return true;
  }
  // suggestion api
  // targets apps that provision and provide internet-capable configurations.
  private fun connectToWifiB(ssid: String, password: String): Boolean {
  
    val suggestion = WifiNetworkSuggestion.Builder()
        .setSsid(ssid)
        .setWpa2Passphrase(password)
        .build();

    val suggestions = listOf(suggestion);
    val wifiManager = context.getSystemService(Context.WIFI_SERVICE) as WifiManager;
    val status = wifiManager.addNetworkSuggestions(suggestions);
    
    val panelIntent = Intent(Panel.ACTION_WIFI)
    startActivityForResult(panelIntent, 545)
    return status == WifiManager.STATUS_NETWORK_SUGGESTIONS_SUCCESS
    /* 
    val intentFilter = IntentFilter(WifiManager.ACTION_WIFI_NETWORK_SUGGESTION_POST_CONNECTION);

    val broadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (!intent.action.equals(WifiManager.ACTION_WIFI_NETWORK_SUGGESTION_POST_CONNECTION)) {
                return;
            }
            // do post connect processing here
        }
    };
    context.registerReceiver(broadcastReceiver, intentFilter);
    
    
    return true;
    */
  }
}
```

**Comment by Marwennnne** [comment]
@tmtong I tried your code and I didn't work for me :/ Is there something you can help me with ?

**Comment by Mkohm** [comment]
For connecting to a access point with internet this seems to work fine for me as a workaround:
```
bool success = (Platform.isAndroid)
                        ? await WiFiForIoTPlugin.registerWifiNetwork(wifiName!,
                            password: wifiPassword!,
                            security: NetworkSecurity.WPA)

                        // The connect method does not work on Android, which is a known issue with the plugin, however doing it like this seems to work fine for both platforms.
                        : await WiFiForIoTPlugin.connect(wifiName!,
                            password: wifiPassword!,
                            withInternet: true,
                            joinOnce: false,
                            security: NetworkSecurity.WPA);

```
